src/data/transform.ts

```diff
@@ -435,124 +427,159 @@ export const convertToExcalidrawElements = (
   if (!elements) {
     return [];
   }
+
   const elementStore = new ElementStore();
-  // Push all elements to array as there could be cases where element id
-  // is referenced before element is created
-  elements.forEach((element) => {
-    let elementId = element.id || regenerateId(null);
+
+  // ensure unique ids
+  // ---------------------------------------------------------------------------
+  const ids = new Set<ExcalidrawElement["id"]>();
+  const elementsWithIds = elements.map((element) => {
+    let id = element.id || regenerateId(null);
 
     // To make sure every element has a unique id since regenerateId appends
     // _copy to the original id and if it exists we need to generate again
     // hence a loop
-    while (elementStore.hasElementWithId(elementId)) {
-      elementId = regenerateId(elementId);
+    while (ids.has(id)) {
+      id = regenerateId(id);
     }
-    const elementWithId = { ...element, id: elementId };
-    elementStore.add(elementWithId as ExcalidrawElement);
```

Comment:
why was this removed ? This is needed to preserve ordering of elements so all ids are added first, the current refactor breaks the cases where element id is referenced before creation [ The last testcase in PR description]